attribution/experiment_logger.py

```diff
@@ -227,6 +271,10 @@ def print_attribution_matrix(
             matrix.index = input_tokens_with_pos
             matrix.columns = output_tokens_with_pos
 
+            if is_hierarchical:
+                # Output is a saliency map which is depth dependent, so normalize the values
+                matrix = matrix / matrix.abs().sum().replace(0, 1)
```

Comment:
1. If we're applying normalisation in general, i think it's better done elsewhere, (i.e. when the scores are calculated) so that the actual values in the df are normalised. That way all display methods will have the same values. RN I think we get different values if we look at the matrix vs the df?

2. N.B. We don't necessarily want to min max normalise, because this will artificially inflate low attributions. If the model isn't attributing much to any input tokens, that's a potential marker for hallucination – we don't want to obscure it.

3. this should be division by max depth, shouldn't it? Why divide by the absolute sum?

4. More general question: for each output token, the maximum attribution score of any input token should be 1.0, and the minimum - 1, for both attribution methods, right? Because the most change _any_ perturbation can effect on an output probability is to take it from 100% probability to 0%, or vice versa (though in practice the latter doesn't happen).

Reply:
2. Agreed, no min maxing.

4. Yeah that's my understanding as well.

Reply:
Maybe taking the mean of each tokens scores wouldn't be the end of the world actually. I might be overthinking it.

Reply:
Removed these lines and set default aggregation to mean

Reply:
Yeah, I think mean would be okay, because deeper tokens have higher scores

Reply:
In general, I think we need to go for consistency. RN we have this:
<img width="980" alt="image" src="https://github.com/user-attachments/assets/5bd7c4ac-7342-4366-96e1-2a40cd9535a7">
Per-output-token scores don't match up with total attribution, and it's hard to see the relationship.

Reply:
Oh, yeah that's not right. I'll look into this!

Reply:
I think the case is that attribution matrix was being normalised by the absolute sum and sentence attribution wasn't! Now all aggregation is done within pandas, no custom normalisation, so the sentence attribution for each input token is the mean of the output token attributions:

![image](https://github.com/user-attachments/assets/657a648b-6fad-42b7-bd5a-1298defe554b)

This is also the case if you use a different aggregation, e.g. using "sum" gives:

![image](https://github.com/user-attachments/assets/d39cd948-d58e-4c5a-97fb-872e436c0ead)